### PR TITLE
nicotine-plus: 1.4.1 -> unstable-2020-05-24

### PR DESCRIPTION
**Pull request by emilazy**
###### Motivation for this change

Includes port to Python 3 and GTK+ 3/PyGObject; see https://github.com/Nicotine-Plus/nicotine-plus/pull/106 for details.

This unstable version identifies itself as 1.4.3, and is planned to be released as 2.0.0 after it has had more fixes and testing through the official upstream PPA. The codebase is old and flaky to begin with, and there have been several unrelated bug fixes in the years since the last stable release, so I think it's worth updating to be able to use non-obsolescent versions of the core dependencies, and have nixos-unstable users help shake out any remaining issues with the port.  I verified that searching and downloading seem to work fine.

Drop the rename of the executable to `nicotine-plus`, as the test depends on the executable name and there doesn't seem to be any strong reason to diverge from upstream here (the original Nicotine is long dead). This is technically a breaking change, so it might be best to include a symlink at `$out/bin/nicotine-plus` temporarily.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
Upstream [doesn't want this packaged yet](https://github.com/Nicotine-Plus/nicotine-plus/pull/106#issuecomment-623198530).

**Comment by emilazy** [comment]
Updated; the integration tests now run successfully. Marking as ready for review per https://github.com/Nicotine-Plus/nicotine-plus/issues/99#issuecomment-630346306 :)

**Review by klntsky — APPROVED** [review]

**Comment by emilazy** [comment]
@ofborg build nicotine-plus

**Comment by emilazy** [comment]
Simplified with `strictDeps = false` at the suggestion of @jtojnar.

@Ericson2314 I guess with gobject-introspection, Python, and `checkPhase` this is already a mess for cross-compilation, but disabling `strictDeps` makes me a little anxious; is there a better way to get setup hooks running for `(install)CheckPhase`? [Previously](https://github.com/NixOS/nixpkgs/commit/6bdf21564e74691a5e679f668f63f31f46375c58#diff-18f751e415743dd11cf6c52ba69c34ffR56-R83) none of the setup hooks for `gdk-pixbuf` and similar were running.

**Comment by emilazy** [comment]
Bumped upstream commit; it now uses the gobject-introspection libnotify bindings.

**Comment by emilazy** [comment]
Pushed with some cleanups and optimistically enabling `xvfb_run` on Darwin in the hopes that that will make the build and tests work (`xorg.xorgserver` is supported there, after all). Let's see what ofborg thinks of it.

**Comment by emilazy** [comment]
This now builds on Darwin, though I can't test it. The `xvfb_run` portability change is no longer required for that, as we can't run the integration tests anyway due to `gtk3` not supporting X11 by default on Darwin, so I can drop it if it's undesired.

**Comment by emilazy** [comment]
Updated with new upstream commits, some licensing pedantry, and a fix to the installation code that makes the desktop entry and manual page work.

**Comment by mathiascode** [comment]
@emilazy Some column counts were outdated in the config file. https://github.com/Nicotine-Plus/nicotine-plus/pull/134 should take care of the issue.

**Comment by mathiascode** [comment]
Nicotine+ 2.0.0 has been released.

**Comment by mathiascode** [comment]
Nicotine+ 2.0.1 has been released.

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 1]
```diff
@@ -1,44 +1,116 @@
...
+, geoip
 
-with stdenv.lib;
+# Test dependencies
+, xvfb_run
+, gnome3
+}:
 
-python27Packages.buildPythonApplication {
+python3.pkgs.buildPythonApplication {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:16]:
```suggestion
python3Packages.buildPythonApplication {
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 2]
```diff
@@ -1,44 +1,116 @@
...
-  version = "1.4.1";
+  version = "unstable-2020-05-24";
 
   src = fetchFromGitHub {
     owner = "Nicotine-Plus";
     repo = "nicotine-plus";
-    rev = "4e057d64184885c63488d4213ade3233bd33e67b";
-    sha256 = "11j2qm67sszfqq730czsr2zmpgkghsb50556ax1vlpm7rw3gm33c";
+    rev = "0f9e4e1c2391196d4070df5851ff48601aceaf7f";
+    hash = "sha256-kSDqoxmX9beqL98tMQylbhCZVMmQy6/oVOyW1xnxFLY=";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:24]:
```suggestion
    sha256 = "sha256-kSDqoxmX9beqL98tMQylbhCZVMmQy6/oVOyW1xnxFLY=";
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 3]
```diff
@@ -1,44 +1,116 @@
...
+    pygobject3
     miniupnpc
     mutagen
-    notify
     (GeoIP.override { inherit geoip; })
   ];
 
-  # Insert real docs directory.
-  # os.getcwd() is not needed
+  checkInputs = [ python3.pkgs.pytest ] ++ lib.optionals (!stdenv.isDarwin) [
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:46]:
```suggestion
  checkInputs = [ python3Packages.pytest ] ++ lib.optionals (!stdenv.isDarwin) [
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 4]
```diff
@@ -1,44 +1,116 @@
...
+    # The gtk3 package doesn't enable X11 on Darwin, so we can't
+    # use xvfb-run.
+    xvfb_run
+    python3.pkgs.robotframework
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:50]:
```suggestion
    python3Packages.robotframework
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 5]
```diff
@@ -1,44 +1,116 @@
...
+
+  # Run setup hooks for installCheckPhase;
+  # see https://github.com/NixOS/nixpkgs/issues/56943.
+  strictDeps = false;
+
   postPatch = ''
-    substituteInPlace ./pynicotine/gtkgui/frame.py \
-      --replace "paths.append(os.getcwd())" "paths.append('"$out"/doc')"
+    # Remove non-free files.
+    sh debian/nicotine-rm-nonfree
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:61]:
```suggestion
    sh debian/nicotine-rm-nonfree
```

Please reference bash from nixpkgs.

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 6]
```diff
@@ -1,44 +1,116 @@
...
+  '';
+
+  installCheckPhase = ''
+    runHook preInstallCheck
+
+    export PATH="$out/bin:$PATH"
+    export HOME="$(mktemp -d)"
+
+    # These tests should be kept in sync with debian/tests/control to
+    # the extent possible.
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:90]:
```suggestion
    # These tests should be kept in sync with debian/tests/control to the extent possible.
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 7]
```diff
@@ -1,44 +1,116 @@
...
+
+    nicotine --version | grep -q 'Nicotine+'
+
+    pytest test/unit
+
+  '' + lib.optionalString (!stdenv.isDarwin) ''
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:96]:
```suggestion
    nicotine --version | grep -q 'Nicotine+'
    pytest test/unit
  '' + lib.optionalString (!stdenv.isDarwin) ''
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 8]
```diff
@@ -1,44 +1,116 @@
...
+    xvfb-run -s '-screen 0 1024x768x24' \
+      robot test/integration/nicotine.robot
+
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:99]:
```suggestion
      robot test/integration/nicotine.robot
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 9]
```diff
@@ -1,44 +1,116 @@
...
+  '' + ''
+    runHook postInstallCheck
   '';
 
   meta = {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:104]:
```suggestion
  meta = lib; {
```

File: pkgs/applications/networking/soulseek/nicotine-plus/default.nix [hunk 10]
```diff
@@ -1,44 +1,116 @@
...
-    maintainers = with maintainers; [ klntsky ];
-    platforms = platforms.unix;
+    license = [
+      # Main license
+      lib.licenses.gpl3Plus
+      # Per sounds/default/license.txt
+      lib.licenses.cc0
+    ];
+    maintainers = with lib.maintainers; [ klntsky emily ];
+    platforms = lib.platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/soulseek/nicotine-plus/default.nix:114]:
```suggestion
      licenses.gpl3Plus
      # Per sounds/default/license.txt
      licenses.cc0
    ];
    maintainers = with maintainers; [ klntsky emily ];
    platforms = platforms.unix;
```

File: pkgs/development/python-modules/miniupnpc/default.nix [hunk 11]
```diff
@@ -1,18 +1,14 @@
...
-  version = "2.0.2";
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "0ca94zz7sr2x57j218aypxqcwkr23n8js30f3yrvvqbg929nr93y";
-  };
+  inherit (miniupnpc_2) version src nativeBuildInputs patches doCheck;
 
-  meta = with stdenv.lib; {
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/miniupnpc/default.nix:8]:
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/miniupnpc/default.nix [hunk 12]
```diff
@@ -1,18 +1,14 @@
...
     description = "miniUPnP client";
     homepage = "http://miniupnp.free.fr/";
-    license = licenses.mit;
-    maintainers = with maintainers; [ peterhoeg ];
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ peterhoeg ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/miniupnpc/default.nix:12]:
```suggestion
    license = licenses.mit;
    maintainers = ib.maintainers; [ peterhoeg ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by peterhoeg** [comment]
We already have 3.0.6 in master.

**Pull request closed without merging**